### PR TITLE
FIX: make un-used ticks not be visible

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Addresses #10874 

This code fails master:

```python
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot(111, frameon=True)
ax.plot()

plt.get_current_fig_manager().canvas.draw()

for xtick in ax.get_xticklabels():
    xtick.get_window_extent()
```

This was originally caused by #9452 adding string data to the tick labels that were not seen.  Previously `get_window_extent` would bail on an empty string.  But after #9452 the labels were not empty.  But they were never drawn, so they never had `self._renderer` assigned.  This makes the labels to be visible if they are not visible, and that causes `get_window_extent` to ignore them.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
I don't think this is 2.2.3 material, but others may disagree...

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1102,7 +1102,12 @@ def _update_ticks(self, renderer):
             tick.set_label1(label)
             tick.set_label2(label)
             if not mtransforms.interval_contains(interval_expanded, loc):
+                tick.set_visible(False)
+                tick.label1.set_visible(False)
```
Comment (anntzer) [inline lib/matplotlib/axis.py]: either also mark the gridline and ticks themselves as invisible, or add a comment about what's happening

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1102,7 +1102,12 @@ def _update_ticks(self, renderer):
             tick.set_label1(label)
             tick.set_label2(label)
             if not mtransforms.interval_contains(interval_expanded, loc):
+                tick.set_visible(False)
```
Comment (anntzer) [inline lib/matplotlib/axis.py]: does this line have any effect?

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/axis.py]: I have written in many places that we should be stricter and require the ticker to only return visible ticks (after all it has all the info needed to do so).  But I guess that's not happening any time soon and instead we just move the issues further down the stack :/

Comment (jklymak) [reply lib/matplotlib/axis.py]: Agenda item for today’s call!

**Comment by dstansby** [comment]
I've put this to `2.2.3`, since it's a regression in the `2.2` branch.

**Comment by efiring** [comment]
This is all very confusing, and it seems like a case of whack-a-mole.  Isn't there a cleaner solution than putting invisible ticks into a list of ticks to draw?  I think @aantzer is pointing to it in his comment https://github.com/matplotlib/matplotlib/issues/9397#issuecomment-336596634.  I suspect the root issue is the interaction between the Locator and autoscaling, and what is needed is an earlier distinction between potential tick locations, relevant to autoscaling, and actual tick locations that will be drawn in the end.  This is further complicated by the usual floating point fudge to include a tick that falls an infinitessimal distance beyond the edge.

**Comment by efiring** [comment]
To consolidate: the moles include #10874, #9397, and #9452.
This PR might be OK as a bugfix for 2.2.3--one last whack, hope that mole is knocked out cold--but I would not like to see it as the real solution for 3.0.

**Comment by jklymak** [comment]
Sorry. Thought I’d opened an issue for this.  Yes ticks that are returned from the locators should just be the ones thatll be drawn not a couple of extra in case of round off errors.  This fix is clearly just a bandaid

**Comment by jklymak** [comment]
Ah, right.  See #5665 for recent discussion with @timhoffm planning to take the lead on improving ticks.

**Comment by jklymak** [comment]
Closing in lieu of #11004, though feel free to resurrect if we want in 2.2.3

**Pull request closed without merging**